src/modules/dashboard/sagas/roles.ts

```diff
@@ -54,6 +62,82 @@ function* colonyDomainUserRolesFetch({
   return null;
 }
 
+const getRoleSetFunctionName = role => {
+  switch (role) {
+    case COLONY_ROLE_ADMINISTRATION:
+      return 'setAdministrationRole';
+
+    // case COLONY_ROLE_ARBITRATION:
+    //   return 'setArbitrationRole';
+
+    case COLONY_ROLE_ARCHITECTURE:
+    case COLONY_ROLE_ARCHITECTURE_SUBDOMAIN:
+      return 'setArchitectureRole';
+
+    case COLONY_ROLE_FUNDING:
+      return 'setFundingRole';
+
+    case COLONY_ROLE_RECOVERY:
+      return 'setRecoveryRole';
+
+    case COLONY_ROLE_ROOT:
+      return 'setRootRole';
+
+    default:
+      throw new Error(`Invalid role to set: ${role}`);
+  }
+};
+
+function* colonyDomainUserRolesSet({
+  payload: { colonyAddress, domainId, userAddress, roles },
+  meta,
+}: Action<typeof ActionTypes.COLONY_DOMAIN_USER_ROLES_SET>) {
+  try {
+    const existingRoles = yield* executeQuery(getColonyDomainUserRoles, {
+      metadata: { colonyAddress },
+      args: { domainId, userAddress },
+    });
+
+    const toChange = Object.keys(roles).reduce(
+      (acc, role) =>
+        existingRoles[role] !== roles[role]
+          ? [...acc, [role, roles[role]]]
+          : acc,
+      [],
+    );
+
+    yield all(
+      toChange.map(([role, setTo], index) =>
+        fork(createTransaction, `${meta.id}_${getRoleSetFunctionName(role)}`, {
+          context: COLONY_CONTEXT,
+          identifier: colonyAddress,
+          methodName: getRoleSetFunctionName(role),
+          params: { address: userAddress, domainId, setTo },
+          ready: true,
+          group: {
+            key: 'setRoles',
+            id: meta.id,
+            index,
+          },
+        }),
+      ),
+    );
+
+    yield put<Action<typeof ActionTypes.COLONY_DOMAIN_USER_ROLES_SET_SUCCESS>>({
```

Comment:
Likewise, you can remove `typeof`. In fact, it's also possible to do `yield put<AllActions>` and it's smart enough to work it which action by the type.